### PR TITLE
ROS2: Image proc non zero crop

### PR DESCRIPTION
**Pull request by klintan**
As per issue https://github.com/ros-perception/image_pipeline/issues/430 there are some missing nodes from image_proc package that needs to be added.

This is for the CropNonZeroNode.

It is built very very similar to the Rectify and Debayer node, so maybe review should be conjunction with https://github.com/ros-perception/image_pipeline/pull/457 .

### CODE REVIEW

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_non_zero.cpp]: commented out?

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_non_zero.cpp]: Can you change this all to 2 spaces per indent?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
LGTM, is this just a straight port? Anything you want to mention?

File: image_proc/src/crop_non_zero.cpp
Comment (klintan) [reply image_proc/src/crop_non_zero.cpp]: Will fix!

**Comment by klintan** [comment]
> LGTM, is this just a straight port? Anything you want to mention?

Nothing in particular, I need to do some more testing. But should be a straight port.

**Comment by SteveMacenski** [comment]
OK tell me when you're happy with it after testing all the relevent options

File: image_proc/src/crop_non_zero.cpp
Comment (klintan) [reply image_proc/src/crop_non_zero.cpp]: Fixed everything to be 2 spaces indentation.

File: image_proc/src/crop_non_zero.cpp
Comment (klintan) [reply image_proc/src/crop_non_zero.cpp]: Used this and remvoved the other since this is the valid one for c++11 or newer.

**Comment by klintan** [comment]
Rebased on top of the new master and fixed PR feedback. Fixed some other issues I forgot to handle when testing it. I couldn't find any issues now at least, after testing. 

Let me know if I missed something or if you have any feedback on my changes.

File: image_proc/CMakeLists.txt
Comment (SteveMacenski) [inline image_proc/CMakeLists.txt:99]: Har har har, `${dependencies}` variable comes back :-)

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Small stuff but looks fine

**Comment by klintan** [comment]
Tests failing on  `image_publisher` not sure why, perhaps a temporary Circle issue, I don't have rights to rerun the workflow either. Should not affect this code.

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_non_zero.cpp]: Commented out?

File: image_proc/include/crop_non_zero.hpp
Comment (SteveMacenski) [inline image_proc/include/crop_non_zero.hpp]: Can you rename image_pub / sub? I assumed those were actually pub/subs and not the string names. That offered a bit of confusion

File: image_proc/include/crop_non_zero.hpp
Comment (SteveMacenski) [inline image_proc/include/crop_non_zero.hpp]: "raw"?

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [reply image_proc/src/crop_non_zero.cpp]: You use this below

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_non_zero.cpp]: Should the type of compression be a parameter? (actually asking, I'm not sure what the ROS1 version did)

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_non_zero.cpp:78]: So connectCb is only called in this lambda, and the next line in the lambda relocks the same mutex. It would be better to either:

a) put the connectCb stuff  inside of the lambda, if there's nothing being otherwise exposed from the outside that needs it
b) only have the mutex in the lambda before the connectCb is called so we're not having 2 locks toggling quickly

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: image_proc/src/crop_non_zero.cpp
Comment (klintan) [reply image_proc/src/crop_non_zero.cpp]: thanks, forgot to uncomment when I switched to this other version requiring this one. Not sure why the compiler did not complain.

File: image_proc/include/crop_non_zero.hpp
Comment (klintan) [reply image_proc/include/crop_non_zero.hpp]: Because of the compression type (also the name ROS1 uses). But I'm open to changing.

File: image_proc/src/crop_non_zero.cpp
Comment (klintan) [reply image_proc/src/crop_non_zero.cpp]: Seems it's just hard-coded: 

```    
image_transport::TransportHints hints("raw", ros::TransportHints(), getPrivateNodeHandle());

sub_raw_ = it_->subscribe("image_raw", 1, &CropNonZeroNodelet::imageCb, this, hints);
```

File: image_proc/src/crop_non_zero.cpp
Comment (klintan) [reply image_proc/src/crop_non_zero.cpp:78]: Went for b) (thinking there was some logic for remove a subscription when the subscriber number were 0 that might be added in the future)

File: image_proc/include/crop_non_zero.hpp
Comment (SteveMacenski) [reply image_proc/include/crop_non_zero.hpp]: Ah ok. Thats fine

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [reply image_proc/src/crop_non_zero.cpp]: Eh. Not ideal but not going to make it part of this PR

**Review by SteveMacenski — APPROVED** [review]
LGTM 

just needs circle ci building

**Comment by SteveMacenski** [comment]
@klintan pull in ros2 once I have #470 merged. This will fix builds so it should just be whatever changes you've made

File: image_proc/include/crop_non_zero.hpp
Comment (SteveMacenski) [inline image_proc/include/crop_non_zero.hpp]: You need to update your guards

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by JWhitleyWork** [comment]
CI restarted. The error that came up is usually because the compiler ran out of memory.

File: image_proc/include/crop_non_zero.hpp
Comment (klintan) [reply image_proc/include/crop_non_zero.hpp]: Done

**Comment by klintan** [comment]
> @klintan pull in ros2 once I have #470 merged. This will fix builds so it should just be whatever changes you've made

Hm I did a rebase on the latest changes, but still failing :/

**Comment by JWhitleyWork** [comment]
@klintan - I'm testing this locally to see what's up. It's a compiler error but I'm not sure why it's happening yet.

**Comment by JWhitleyWork** [comment]
@klintan - I created https://github.com/klintan/image_pipeline/pull/3, which I think will solve the problem. Please test it.

**Comment by SteveMacenski** [comment]
@JWhitleyAStuff can you PR that and just merge it? That's been irritating me as well

**Review by SteveMacenski — APPROVED** [review]
I'm happy again when @JWhitleyAStuff is

**Comment by JWhitleyWork** [comment]
@klintan - Once you rebase on `master`, I'm good to merge.

**Comment by SteveMacenski** [comment]
You mean on `ros2`, right?

**Comment by JWhitleyWork** [comment]
Yes, sorry.

**Comment by klintan** [comment]
i'll rebase asap.

**Comment by klintan** [comment]
@JWhitleyAStuff I had to change 
`dst.create(src.rows / 2, src.cols / 2, cv::DataType<DstPixel>::type);`
to
`dst.create(src.rows / 2, src.cols / 2, cv::traits::Type<DstPixel>::value);` 
per https://github.com/opencv/opencv/issues/10115 , which version of opencv do you run? (I'm on 3.4.5) otherwise it wouldn't build. Perhaps it's something with my env since circleci seems to be fine with that ? 

Fixed all other issues.  Perhaps the crop_decimate fix should be in a different PR if it's a problem.

edit: circleci does not build with that change, so reverted regardless. But I think it's worth looking into that issue.

**Comment by SteveMacenski** [comment]
@klintan that's a good use for the preprocessor `if`s on the version, there's a bunch of examples in the code for supporting multiple opencv versions that way

**Comment by klintan** [comment]
> @klintan that's a good use for the preprocessor `if`s on the version, there's a bunch of examples in the code for supporting multiple opencv versions that way


Ok used a condition for the opencv version, that fixed the issue on circle and locally.

File: image_proc/src/crop_decimate.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_decimate.cpp:49]: Does the minor version actually matter? just curious.

**Comment by SteveMacenski** [comment]
Just waiting on @JWhitleyAStuff feedback then

File: image_proc/src/crop_decimate.cpp
Comment (klintan) [reply image_proc/src/crop_decimate.cpp:49]: Yes, or put another way; the change was made in 3.3.1 it seems, https://github.com/ros-perception/image_pipeline/pull/302

edit: to be safe, I guess we would need a CV_SUBMINOR_VERSION as well ? 3.3.0 would use the new way as well. What do you think?

**Review by JWhitleyWork — APPROVED** [review]
LGTM!

File: image_proc/src/crop_decimate.cpp
Comment (SteveMacenski) [reply image_proc/src/crop_decimate.cpp:49]: I mean I think this is fine

**Comment by SteveMacenski** [comment]
@klintan any thing else you want to add before merging?

**Pull request merged**
